### PR TITLE
feat: add LumberjackLogFactory

### PR DESCRIPTION
**Pull request by LayZeeDK**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Using `LumberjackLogBuilder` requires passing `LumberjackTimeService` to maintain testability. This can be cumbersome.

Issue Number: N/A

## What is the new behavior?
`LumberjackLogFactory` removes the need to having to inject and pass `LumberjackTimeService`.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: libs/ngworker/lumberjack/src/lib/logging/lumberjack-log-factory.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import { LumberjackLevel } from '../logs/lumberjack-level';
+import { LumberjackLogPayload } from '../logs/lumberjack-log-payload';
+import { LumberjackTimeService } from '../time/lumberjack-time.service';
+
+import { LumberjackLogBuilder } from './lumberjack-log.builder';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class LumberjackLogFactory<TPayload extends LumberjackLogPayload | void = void> {
```
Comment (NachoVazquez) [inline libs/ngworker/lumberjack/src/lib/logging/lumberjack-log-factory.ts:12]: Should we document it a bit?

**Review by NachoVazquez — CHANGES_REQUESTED** [review]
It makes the test much cleaner.

I missed a test file for the factory.

File: libs/ngworker/lumberjack/src/lib/logging/lumberjack-log-factory.ts (hunk 1 above)
Comment (LayZeeDK) [reply libs/ngworker/lumberjack/src/lib/logging/lumberjack-log-factory.ts:12]: For sure! But I wanted your feedback on the validity of this idea first 😃

**Comment by LayZeeDK** [comment]
> I missed a test file for the factory.
Indeed. I wanted your feedback on this concept first 🙂

**Comment by LayZeeDK** [comment]
I added inline docs. Let's add docs in README in the docs review branch after merging.

File: libs/ngworker/lumberjack/src/lib/logging/lumberjack-log-factory.spec.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+      expect(log.payload?.test).toBe(true);
+    });
+
+    it('creates a log with a dynamic payload', () => {
+      const log = logFactoryWithPayload.createInfoLog(testMessage).build({ test: true });
+
+      expect(log.payload?.test).toBe(true);
+    });
+  });
```
Comment (NachoVazquez) [inline libs/ngworker/lumberjack/src/lib/logging/lumberjack-log-factory.spec.ts:101]: Maybe a check that the payload is undefined when not passed.
```suggestion
  describe('Payload', () => {
    interface TestPayload extends LumberjackLogPayload {
      readonly test: boolean;
    }

    beforeEach(() => {
      logFactoryWithPayload = resolveDependency(LumberjackLogFactory);
      logFactoryWithoutPayload = resolveDependency(LumberjackLogFactory);
    });

    let logFactoryWithPayload: LumberjackLogFactory<TestPayload>;
    let logFactoryWithoutPayload: LumberjackLogFactory;

    it('creates a log with a static payload', () => {
      const log = logFactoryWithPayload.createErrorLog(testMessage).withPayload({ test: true }).build();

      expect(log.payload?.test).toBe(true);
    });

    it('creates a log with a dynamic payload', () => {
      const log = logFactoryWithPayload.createInfoLog(testMessage).build({ test: true });

      expect(log.payload?.test).toBe(true);
    });

    it('creates a log without a payload', () => {
      const log = logFactoryWithoutPayload.createInfoLog(testMessage).build();

      expect(log.payload).toBeUndefined();
    });
  });
    
     it('creates a log without payload', () => {
      const log = logFactoryWithPayload.createInfoLog(testMessage).build();

      expect(log.payload?.test).toBe(true);
    });
  });
```

**Review by NachoVazquez — CHANGES_REQUESTED** [review]

**Review by NachoVazquez — APPROVED** [review]

**Pull request merged**